docs/design/sql/14-automatic-query-resubmission.md

```diff
@@ -0,0 +1,157 @@
+|Related Github issues|https://github.com/hazelcast/hazelcast/pull/21635|
+|Document Status / Completeness| Complete                                          |
+|Requirement owner| Sandeep Akhouri                                              |
+|TDD Author|Viliam Ďurina|
+|Developer(s)| Krzysztof Ślusarski |
+|Quality Engineer| TODO                             |
+|Support Engineer| TODO                             |
+|Technical Reviewers| TODO                          |
+|Simulator or Soak Test PR(s) | TODO            |
+
+### Background
+#### Description
+
+A customer perceives that AP is NOT implemented correctly in Hazelcast. If the
+member hosting the primary partition is NOT available, their expectation is that
+data should be transparently served from the backup. When the cluster is
+changing topology, e.g., cluster member that contains the primary data is
+killed, the client throws errors.
+
+Automatic retries, with an upper bound on the number of retries, are implemented
+with many IMap operations on the client. However, with SQL, the client cannot do
+this, because it doesn't understand the query, and it doesn't know if the query
+is side-effect-free and it's safe to be retried. Secondly, the client might
+already have received some result rows, and if it retries the query, those rows
+will be received again.
+
+#### Scope
+
+The scope is to add support for automatic SQL retries in Java client. Other
+clients might follow, but it's not in the scope of the initial PR, as the
+requirement is currently only for Java client.
+
+There will be no retry mechanism for queries submitted through member instances.
+
+#### Terminology
+
+| Term           | Definition                            |
+|----------------|---------------------------------------|
+| Topology error | Query failure due to a member failure |
+
+### Technical Design
+
+The client doesn't understand the SQL statements it sends, and we want it to
+remain this way in order to simplify the implementation in multiple languages.
+We also don't want to retry statements by default, as not all are read-only.
+
+We will analyze three scenarios:
+1. Some, but not all rows were already received when the query fails
+2. No rows were received yet, the query text starts with `SELECT` (case-insensitive, ignoring 
+       white space) 
+3. No rows were received yet, the query _doesn't_ start with `SELECT`
+
+#### Resubmission policies
+
+We propose that the user will be able to pick one of these retry policies:
+
+1. `NEVER`: the current state. If a query fails, the failure is immediately 
+   forwarded to the user
+2. `RETRY_SELECTS`: the query will be retried if:
+   - no rows were received yet
+   - the SQL text starts with `SELECT` (case-insensitive, ignoring
+     white space)
+3. `RETRY_SELECTS_ALLOW_DUPLICATES`: as before, but query can be retried
+   after it returned some rows. The already-returned rows will be returned again.
```

Comment:
What will happen to the previous started sql job? Will it be cancelled? Can you please write to the TDD?

Reply:
Similar as above - since we retry only after either an error from a member (meaning the member already handled the error), or after a connection loss (meaning the member won't be able to send the response), there's no need to cancel.